sdk/python/tests/integration/feature_repos/test_repo_configuration.py

```diff
@@ -71,91 +73,140 @@ def ds_creator_path(cls: str):
 ]
 
 
-OFFLINE_STORES: List[str] = []
-ONLINE_STORES: List[str] = []
-PROVIDERS: List[str] = []
+def construct_universal_entities() -> Dict[str, List[Any]]:
+    return {"customer": list(range(1001, 1110)), "driver": list(range(5001, 5110))}
+
+
+def construct_universal_datasets(
+    entities: Dict[str, List[Any]], start_time: datetime, end_time: datetime
+) -> Dict[str, pd.DataFrame]:
+    customer_df = driver_test_data.create_customer_daily_profile_df(
+        entities["customer"], start_time, end_time
+    )
+    driver_df = driver_test_data.create_driver_hourly_stats_df(
+        entities["driver"], start_time, end_time
+    )
+    orders_df = driver_test_data.create_orders_df(
+        customers=entities["customer"],
+        drivers=entities["driver"],
+        start_date=end_time - timedelta(days=365),
+        end_date=end_time + timedelta(days=365),
+        order_count=1000,
+    )
+
+    return {"customer": customer_df, "driver": driver_df, "orders": orders_df}
+
+
+def construct_universal_data_sources(
+    datasets: Dict[str, pd.DataFrame], data_source_creator: DataSourceCreator
+) -> Dict[str, DataSource]:
+    customer_ds = data_source_creator.create_data_source(
+        datasets["customer"],
+        destination="customer_profile",
+        event_timestamp_column="event_timestamp",
+        created_timestamp_column="created",
+    )
+    driver_ds = data_source_creator.create_data_source(
+        datasets["driver"],
+        destination="driver_hourly",
+        event_timestamp_column="event_timestamp",
+        created_timestamp_column="created",
+    )
+    orders_ds = data_source_creator.create_data_source(
+        datasets["orders"],
+        destination="orders",
+        event_timestamp_column="event_timestamp",
+        created_timestamp_column="created",
+    )
+    return {"customer": customer_ds, "driver": driver_ds, "orders": orders_ds}
+
+
+def construct_universal_feature_views(
+    data_sources: Dict[str, DataSource]
+) -> Dict[str, FeatureView]:
+    return {
+        "customer": create_customer_daily_profile_feature_view(
+            data_sources["customer"]
+        ),
+        "driver": create_driver_hourly_stats_feature_view(data_sources["driver"]),
+    }
+
+
+def setup_entities(
+    environment: "Environment", entities_override: Optional[Dict[str, List[Any]]] = None
+) -> "Environment":
+    environment.entities = (
+        entities_override if entities_override else construct_universal_entities()
+    )
+    return environment
+
+
+def setup_datasets(
+    environment: "Environment",
+    datasets_override: Optional[Dict[str, pd.DataFrame]] = None,
+) -> "Environment":
+    environment.datasets = (
+        datasets_override
+        if datasets_override
+        else construct_universal_datasets(
```

Comment:
> if you don't, it uses entities (which might have been overwritten previously) and then you could run into an error if you don't have customer or driver entities.

Yea, but in which case would you override the entities but not override the steps that follow (like `setup_datasets`)? Presumably you as the author would know that there are downstream effects.

I think the real question is what the value of `setup_datasets` is in the first place. It seems like we can get away with having `construct_universal_datasets` which seems useful, and otherwise just letting people operate on the environment object directly with custom code. This function isn't really doing much tbh.

> Could we almost do some sort of like builder approach? where basically nothing is assigned until a .build() method is called. Then we'd have only really 3 usage patterns

Getting some real AutoValue/Lombok vibes here :)

> Environment.builder().setEntities(...).setDatasets(...).setFeatureViews(...).build() -> force the user to explicitly define all the objects if they need something fully custom
How would this be different from a constructor like Environment(entities=.., datasets=...)?

> Environment.getDefault().toBuilder().setEntities(...).build() -> explicitly communicates that there are other parts of the environment that are "default" that may or may not be compatible with the entities

This seems very close to having
```
e = Environment(entities=.., datasets=...)
# My custom code
e.entities = do_my_custom_stuff()
e.setup()
```
To me the main value of the builder pattern in this context is that we can delay the actual execution/staging. I'm not sure if the getDefault/toBuilder logic is as important.